Add draft CapTP spec (#42)
ocapn/ocapn: draft-specifications/CapTP Specification.md

-   [OCapN Locators](): Specification covers representation of object references
    for both in band and out of band usage.

# CapTP Overview

CapTP supports many features, as described in the [Introduction](#introduction)
section. Many different things can happen within a CapTP session, with a session
being two entities communicating over CapTP through a  a reliable, in-order
channel called a OCapN Netlayer. These specifics of the Netlayers are specified
in the OCapN Netlayer specification.

Here's an overview of some things which may occur during a CapTP session:

1. A session is established pairwise between machines.
2. Communication occurs between objects.
    -   Initial connectivity to objects is established by querying the bootstrap
        object for access to known objects (via sturdyrefs or certificates).
    -   Messages are exchanged between objects which hold references to each
        other.
    -   Promises may be created when messages are sent.
    -   Messages may be pipelined to promises, queueing those messages to
        eventually be delivered to their resolution.
    -   Handoffs are initiated when sending a message with a reference to an

>         object on a third CapTP session.

Author: I think third is correct here, however I think christine was also confused by my use of "third" at one of the earlier passes through she did so I'll change this to different wording. I wanted to be very explicit that it's a reference outside of the session in question so I've gone with slightly different wording than you proposed here:

> - Handoffs are initiated when sending a message with a reference to an object outside of the CapTP session.

established and set up, and can be used to send and receive messages with remote
objects.

## Sending and receiving messages

When an object sends a message to another object across a CapTP boundary, it
either expects a reply or not. If the object expects a reply, a
[promise](#promises) is created to represent the reply. To do this, we send a
[`op:deliver`](#op-deliver) message. If the message is delivered without
importing an answer, this is an [`op:deliver-only`](#op-deliver-only) message.
Messages can also be "pipelined" to promises generated by message sends with
`op:deliver`, eliminating the need for unnecessary round-trips via [Promise
Pipelining](#promise-pipelining).

This covers the basic usage of CapTP. Another aspect that can occur when sending
messages through CapTP is when an object sends a message that includes a
reference to an object residing in a different session. To perform this while
maintaining the security CapTP provides, we need to perform a [Third Party
Handoff](#third-party-handoffs).

It is conventional, but not required, that some objects may offer different
behavior depending on a "method". The method is conventionally a symbol whose

> value is the name of the method/behavior. This symbol is the first argument in a

Reviewer: #54

message, and the remaining arguments are the message sent to that behavior. This
convention is followed for the CapTP bootstrap object & promises.

## Ending a session

Finally, a session may end due to an error occuring that could not be continued
from or because either side wishes to end the session. Both situations are
covered by the [`op:abort`](#op-abort) message. When this is received the
session should be severed and unresolved promises broken.

# Promises

Promises are a key part of CapTP. They are used to represent a value which is
not yet known. Promises are created by sending an `op:deliver` message, where a
promise is created for the value of the response. They can be chained together
in what is called [Promise Pipelining](#promise-pipelining), whereby messages

> are sent to the promise which should be delivered if the promise resolves.

Author: I've changed this to fulfill and hopefully made it clearer with the language I've used here.

> ## [Promise and Resolver Objects](#promise-objects)

Author: Yeah. I had wanted a good way to provide links to sections within a document and couldn't find a standard markdown way of doing this. However I did see a few suggestions to make your headings anchor links which you can then refer to later in the document.

Another solution I saw was to "guess" what the slug would be that markdown readers/renders would give it, but I wasn't really keen on this solution (I suspect the slug could change depending markdown renderer).

I'm going to leave the headlines links for now as I've made heavy use of linking to sections within the rest of the document, however I'm open to other solutions (including potentially using a different markup format, if folks prefer).

2.  Send message to car factory promise to make a car
3.  Send message to car promise to drive
4.  Get back driving noise

The other CapTP session, which is making the car factory, car, and finally car
noise, is able deliver those messages immediately to the local objects on their
side as they become fulfilled and send the car noise back at the end.

# [Cryptography](#cryptography)

Each CapTP session has a per session key pair which is used for signing certain
structures for example in the [third party handoffs
section](#third-party-handoffs). These key pair values are generated EdDSA with
a SHA512 hash.

**NOTE:** These representations are considered temporary and we are anticipating
replacing them, probably with record-based representations.

## [Public key](#public-key)

Public keys are formatted based on gcrypt's s-expression format, using EdDSA
public keys and the SHA512 hash algorithm. The EdDSA public keys are based on
the Ed25519 elliptic curve. The public key is formatted as follows:

```text
(public-key (ecc (curve Ed25519) (flags eddsa) (q ...) (s ...)))
```

Everything except the `...` must be verbatim. The `...` is replaced by Binary

> Data typed values representing the public key.

Author: This representation is based off the s-expression format that gcrypt uses, so this exact expression (except with `q` and `s` values filled with the public key data). The expression is actually a Syrup sequence composed of symbols. Spritely is not attached to this format and we probably would be looking to change it within the future.

I think this might be a good candidate for it's own `desc` struct, but I think this could be best discussed within the group to see what works best for everyone here.

> Data typed values representing the public key.

Author: I've improved the text describing the two cryptography values, hopefully it's a bit clearer now.

exported at.

### Receiving
When the `op:pick` message is received, a promise should be exported at the
answer position specified by `new-answer-pos`. The promise should eventually
resolve to the value at the index specified by `selected-value-pos`, in values
provided in the `promise-pos` promise. If the `promise-pos` promise breaks, or
the `selected-value-pos` is out of bounds, the promise should break.

## [`op:abort`](#op-abort)

This is used to abort a CapTP session, when this is sent the connection should
be severed and any per session information (e.g. session key pair, etc.) should
be removed.

The `op:abort` message is:

```text
<op:abort reason>  ; reason: String
```

## [`op:listen`](#op-listen)

This is used to listen to a promise. This is done in order to get notified when
the promise is fulfilled with a value or broken on an error. Please see the
[promises section](#promises) for more information on how promises work.

The `op:listen` message is:

```text
<op:listen to-desc           ; desc:export | desc:answer
           listen-desc       ; desc:import-object
           wants-partial?    ; boolean
```

> The `wants-partial?` flag indicates if a "partial" update should be provided as
> the notification when the promise resolves. If `wants-partial?` is false and the
> promise resolves to another promise, a notification to the `listen-desc` is not
> sent. A notification is only ever sent to the `listen-desc` when the promise
> resolves to a non-promise value, or breaks. If `wants-partial?` is true, a
> notification on any resolution, including a promise. In the case it does notify
> on a promise the listen is considered complete, if future notifications are
> desired, further `op:listen` operations should be sent to the promises.

Reviewer: Grammar suggestions:
```suggestion
The `wants-partial?` flag indicates if a "partial" update should be provided as
the notification when the promise resolves to another promise. If
`wants-partial?` is false, a notification is sent to the `listen-desc` only when
the promise resolves to a non-promise value or breaks (i.e., not when it
resolves to another promise). If `wants-partial?` is true, a notification is
sent for any resolution, including to another promise. Any notification is
considered to conclude the `op:listen` interaction, and if future notifications
are desired (e.g., after a partial notification) then further `op:listen`
operations should be sent.
```

### Sending

`to-desc` MUST be a `desc:export` or `desc:export` which corresponds to a
promise on the remote side.

`listen-desc` MUST be a `desc:import` that is being imported. This will be
invoked when the promise comes to a resolution.

### Receiving

When receiving this message, providing `to-desc` matches a promise exported to
this session, a mechanism MUST be setup to fulfill or break the provided
resolver when a resolution is available to the `listen-desc` object. If a
resolution is already available, the resolver provided in `listen-desc` MUST be
fulfilled or broken.

## [`op:gc-export`](#op-gc-export)

When a reference is given out to another CapTP session, the reference must be

> kept valid until the other session is done with it. This is achieved by

Reviewer: "other session", or other party within a session?

Author: yeah I agree, I've used session in a number of places in ambiguous ways. To me a session consists of two parties speaking CapTP with one another, however I've used it to mean a single party within a session which is confusing. I've fixed this wording here.

Each time the remote session no longer needs the reference, it should send an
`op:gc-export` message with a `wire-delta` that reflects the number of
references it has been given:

The message looks like:

```text
<op:gc-export [export-pos    ; positive integer
               wire-delta]>  ; positive integer
```

## [`op:gc-answer`](#op-gc-answer)

When a [`op:deliver`](#op-deliver) is sent with an `answer-pos` for use with
promise pipelining. The receiver will create a promise at the answer position.
The receiver needs to know when it's able to garbage collect this promise. This
is done by sending an `op:gc-answer` message. The `answer-pos` in this message
MUST correspond to the `answer-pos` in the [`op:deliver`](#op-deliver) message,
that you are no longer interested in.

Once the `answer-pos` has been GC'd through sending the `op:gc-answer`
operation, the `answer-pos` can be re-used.

```text
<op:gc-answer answer-pos>  ; answer-pos: positive integer
```

# Descriptors

Several operations (e.g. `desc:import-object` and `desc:export`) are describing
importing and exporting objects. There had to be a choice if these actions

> should be described from the sender\'s or receiver's side, in this case we

Reviewer: Markdown nit: this backslash should be unnecessary (at least for GitHub)
```suggestion
should be described from the sender's or receiver's side, in this case we
```

> choose the receiver's side. This means if an object is exported from session A
> to session B, session A sends a `desc:import-object` as session A is describing
> it from B's perspective.

Reviewer: **session**, or **party**?
```suggestion
choose the receiver's side. This means if an object is exported from Alice to
Bob, Alice sends a `desc:import-object` to describe it it from Bob's
perspective.
```

## [`desc:import-object`](#desc-import-object)

Any object which is exported over CapTP is described with a positive integer.
This positive integer MUST be unique to this CapTP session and refer to this
specific object.

```text
<desc:import-object position>  ; position: positive integer
```

## [`desc:import-promise`](#desc-import-promise)

When a promise is exported over a CapTP boundary is it described with a
`desc:import-promise` message. This message contains a positive integer which is

> unique to this CapTP session and refers to this specific promise.

Reviewer: Unique to the **session**, or unique to the **party** _within_ the session?

```

## [`desc:export`](#desc-export)

When a message is sent across a CapTP boundary that refers to an imported object
within the session the message is being sent to (either a `desc:import-object`
or a `desc:import-promise`), then this should be referred to with a
`desc:export`. The position MUST be the positive integer provided in the import
descriptor.

If an object reference is being sent from a different session, see the [Third
Party Handoffs](#third-party-handoffs) section.

```text
<desc:export position>  ; position: positive integer
```

## [`desc:answer`](#desc-answer)

This is used to refer to a promise which is being pipelined. The position MUST
be the positive integer provided in the [`op:deliver`](#op-deliver) message.
This should not be referenced after the [`op:gc-answer`](#op-gc-answer) message
has been sent for this position.

```text
<desc:answer answer-pos> ; position: positive integer
```

## [`desc:sig-envelope`](#desc-sig-envelope)

This encapsulates a CapTP object and provides a signature. The signature is
created on the binary data of the serialized CapTP object in the `signed` field.

The process of generating this is:

> 1.  Fully serialize to Syrup octets a CapTP object.

Reviewer: ```suggestion
1.  Fully serialize a CapTP object to a sequence of Syrup octets.
```

### Handoffs from the Exporter's perspective

There are two events which will happen for the exporter, these can happen in any order:

-   The gifter deposits a gift
-   The receiver sends a [`desc:handoff-receive`](#desc-handoff-receive).

The exporter performs the following:

1.  MUST verify the `desc:handoff-receive` (see `desc:handoff-receive` section). If it is incorrect, abort the handoff; otherwise, continue.
3.  If the gift has already been deposited return the gift; otherwise return the gift, when is received.

Note: Gifts are per-session. This means that gifts deposited within the `Gifter <-> Exporter` session should only be available for withdrawal within the `Receiver <-> Exporter session`. The receiver should be the only party able to withdraw the gift left by the gifter. Implementers should ensure that the management of gifts adheres to this per-session requirement, preventing unauthorized access to gifts.

# Operations

## [`op:start-session`](#op-start-session)

> On a new connection a key-pair for this session should be generated. This key-pair should be an EdDSA key-pair with a SHA512 hash.

Reviewer 2: Maybe a comma would help here. "On a new connection, a key-pair..."

The `acceptable-location-sig` MUST be valid that the `session-pubkey` provided a valid signature of `acceptable-location`.

## [`op:bootstrap`](#op-bootstrap)

This operation requests the bootstrap object to be available at the specified `answer-position`. The message looks like:

```text
<op:bootstrap [answer-position    ; positive integer or 0
               resolve-me-desc]>  ; desc:import-object | desc:import-promise
```

### The bootstrap Object

The bootstrap object is responsible for providing access to local objects on the session. It has two different behaviors, these are selected using the conventional CapTP method mechanism of sending a symbol as the first argument, the following methods are available:

-   `fetch`
-   `deposit-gift`
-   `withdraw-gift`

#### `fetch` Method

This method is used to fetch an object from the bootstrap object. To use it you need a `swiss-number` which is a Binary Data type. This swiss number should correspond an object which exists in this session. The result will be the object which corresponds to this `swiss-number` or an error if the object does not exist or a swiss number was not provided.

> Here is an example of how to use this method is:

Reviewer 2: Maybe either "Here is an example of how to use this method" or "An example of how to use this method is:"

This operation delivers a message to an object and does not expect any result from the receiving object.

The `op:deliver-only` message is:

```text
<op:deliver-only [to-desc   ; desc:export
                  args]>    ; Sequence
```

`to-desc` is a `desc:export` descriptor which corresponds to the object the message is being sent to.

`args` is a sequence which are the arguments to invoke the object with.

Since this message has no result, it should be delivered to the object but nothing further is required.

## [`op:deliver`](#op-deliver)

This operation delivers a message to an object with the expectation of a return value which should be installed at the location specified in `answer-pos`. The `op:deliver` message is:

``` text
<op:deliver [to-desc            ; desc:export
             args               ; sequence
             answer-pos         ; positive integer | false
             resolve-me-desc]>  ; desc:import-object | desc:import-promise
```

The `resolve-me-desc` is a `desc:import-object` or `desc:import-promise` which represents a reference to a local object which will be notified upon the resolution of the promise.

### Sending

> When sending a message which expects a result, a promise should be generated on the side which is sending the `op:deliver-only`. This promise should be provided to the object sending this message. See the [promises section](#promises) for more information.

Reviewer 2: Should this be `op:deliver` instead of `op:delivery-only`?

# Status of This Document

This is a draft specification, created by the
[OCapN](https://github.com/ocapn/ocapn) pre-standardization group. This draft is
subject to change with the work of the group; if you're interested in being part
of that work, please join!

Authors: Jessica Tallon, Christine Lemmer-Webber, Mark Miller & The OCapN
Pre-standardization Group.

# [Introduction](#introduction)

Capability Transport Protocol (CapTP) is a secure messaging protocol designed
for communication between objects in distributed systems across networks. By
utilizing a capability security model, CapTP ensures both security and
functionality without compromise. In this model, an object can use a reference
(capability) if (and only if) it possesses a reference to one. In other words,
"if you don't have it, you can't use it."

CapTP offers several valuable features, including:

- Powerful promises and promise pipelining, allowing remote value usage before
  resolution.
- Error propagation across the network.
- Distributed garbage collection.

> - Secure third-party handoffs, even when messages are published.

Author: I've changed this to "Secure third-party handoffs, even when CapTP messages are not kept secret." hopefully that's more clear. It's meaning that the secure third party handoff feature is not dependent on the messages between the three sessions being kept secret. This should work just as well if these messages were public for all to see.

CapTP is built upon the foundation of the actor model, where each actor is

> referred to as an object. An actor model is system where objects pass messages

Reviewer 2: Should this be either "The actor model is a system where..." or "An actor model is a system where..."?

    -   Initial connectivity to objects is established by querying the bootstrap
        object for access to known objects (via sturdyrefs or certificates).
    -   Messages are exchanged between objects which hold references to each
        other.
    -   Promises may be created when messages are sent.
    -   Messages may be pipelined to promises, queueing those messages to
        eventually be delivered to their resolution.
    -   Handoffs are initiated when sending a message with a reference to an
        object on a third CapTP session.
    -   Both parties cooperate to free object references which are recognized to
        no longer be needed to be imported from one side.
5.  A session ends.

## Establishing a connection

A new connection is established either by:

- Using an out-of-band mechanism to connect to a remote session
- Peforming a handoff to connect to a third session

In either situation, a secure channel needs to be created to the remote session.
This is out of scope for this specification but it's covered by the OCapN
Netlayers specification. Once a secure channel is established, we MUST to
perform the following in this order:

> 1.  Create a per-session cryptography key pair (see

Reviewer 3: I wonder why a session key pair is in this layer.

The Agoric CapTP API doesn't use a session key pair.

cc @erights @kriskowal @michaelfig

Reviewer 4: The 3PH presentation from a couple meetings ago goes over how they're used, but I do think it might be good to add some abstraction boundaries; probably most of the crypto should be netlayer specific.

Reviewer 3: I promoted this thread to an issue: #53.

    [Cryptography](#cryptography))
2.  Send a [`op:start-session`](#op-start-session) message
3.  Receive and verify the remote party's
    [`op:start-session`](#op-start-session) (including dialback verification)
5.  Send a [`op:bootstrap`](#op-bootstrap) to import their bootstrap object

(**NOTE:** The `op:bootstrap` is likely to be deprecated soon.)

Once these steps are successfully completed, the connection is considered
established and set up, and can be used to send and receive messages with remote
objects.

## Sending and receiving messages

When an object sends a message to another object across a CapTP boundary, it
either expects a reply or not. If the object expects a reply a
[promise](#promises) is created to represent the reply. To do this, we send a
[`op:deliver`](#op-deliver) message. If the message is delivered without

> importing an answer, this is an [`op:deliver-only`](#op-deliver-only) message.

Reviewer 2: This paragraph uses "a `op:deliver`" and "an `op:deliver-only`". Should these be made consistent?

Reviewer 4: @davexunit, I believe these are two different messages, the latter of which ignores any result.

Reviewer 2: Oh let me be more clear: It's the "a" and "an" part that is inconsistent to me.

behavior depending on a "method". The method is conventionally a symbol whose
value is the name of the method/behavior. This symbol is the first argument in a
message, and the remaining arguments are the message sent to that behavior. This
convention is followed for the CapTP bootstrap object & promises.

## Ending a session

Finally, a session may end due to an error occuring that could not be continued
from or because either side wishes to end the session. Both situations are
covered by the [`op:abort`](#op-abort) message. When this is received the
session should be severed and unresolved promises broken.

# Promises

Promises are a key part of CapTP. They are used to represent a value which is
not yet known. Promises are created by sending an `op:deliver` message, where a
promise is created for the value of the response. They can be chained together
in what is called [Promise Pipelining](#promise-pipelining), whereby messages
are sent to the promise which should be delivered if the promise resolves.

## [Promise and Resolver Objects](#promise-objects)

Promises work like regular objects in CapTP. Promises come as a pair, there is a

> "promise" object which represents a value and may eventually resolve to either a
> value or object reference or may break, either explicitly, implicitly through
> error handling, or via network partition.

Reviewer: ```suggestion
"promise" object which represents a value and may eventually fulfill to either a
value or object reference, or may break by either explicit
instruction or implicit error propagation or network partition.
```

> Messages can be sent to a promise as if it were the resolved object. These sent
> messages will be relayed to the eventual object if it is `fulfill`ed to one.
> There is also a "promise resolver" object which is used to provide the promise a
> value or break on an error. The resolver object has two methods:

Reviewer: ```suggestion
Messages can be sent to a promise as if it were an object. These sent
messages will be relayed to the eventual fulfilled value if it is an object.
There is also a "promise resolver" object which is used to provide the promise a
fulfillment value or break reason. The resolver object has two methods:
```

> - `fulfill`: Provide the promise with its resolved values (success case)

Reviewer: ```suggestion
- `fulfill`: Provide the promise with its fulfillment values (success case)
```

- `break`: Breaks the promise (usually due to an error)

### `fulfill`

This method can take any number of arguments. These arguments can be promise
pipelined on with the [`op:pick`](#op-pick) operator and should be delivered to
any listeners of the promise.

### `break`

> Break takes a single argument, which represents an error that occured. This

Reviewer: This is a surprising asymmetry, which probably merits some explanation.

error should be delivered to any listeners of the promise.

**NOTE:** The value of errors transmitted over CapTP is up to the transmitting
party. However, including the contents of exception objects or debugging
information such as backtraces could unintentionally leak sensitive information.
It may be valuable at a CapTP border to strip certain kinds of debugging
information, to encrypt its contents, or to use the sealer/unsealer pattern from
capability literature to secure its contents so that only parties with the
relevant sealers/unsealers can safely decode them.

## [Promise pipelining](#promise-pipelining)

When an [`op:deliver`](#op-deliver) message is sent and a result is pending, it
is often desirable to use this result. One way to do this would be to wait until
the promise has been resolved and then send the next message to the result.
While this approach can be taken within OCapN, promise pipelining is preferred.
When a message is sent with an `answer-pos` specified, the promise on the remote
session can then be referenced and messaged with the

> [`desc:answer`](#desc-answer) descriptor. When the result is ready, messages
> sent to this promise will have them delivered to the result.

Reviewer: ```suggestion
[`desc:answer`](#desc-answer) descriptor. When a promise is fulfilled
to an object, messages sent to the promise will be delivered to that object.
```

If the answer contains multiple values, the [`op:pick`](#op-pick) operator
MUST be used to choose the value to be pipelined on. This works by exporting the
individual value in another answer position, which then can be used to message
the promise. If multiple values are pipelined on, without the pick operation,

> the resulting promise MUST be broken with an error and not fulfilled.

Reviewer: So there are protocol-level affordances for fulfilling a promise to multiple values (and for automatically breaking messages that do not target a specific member of such a collection)? Interesting.

Author: Yeah. If the promise only represents one value you of course don't need to use `op:pick`, it's just for those promises which represent multiple values. It used to be that promises could only represent a single value, and so we often encoded multiple values in lists, which was fine, but it prevented promise pipelining and caused a lot of unpacking in our code. We solved this by adding this operator and having it so promises could represent multiple values.

has been resolved. Messages can just be sent in advance to a promise of the
eventual object as they normally would to its eventual resolution. This also
improves efficiency by reducing the number of round trips needed to perform the
same task. Take the above example of a car factory which produces a car, the car
then can be driven which will produce a noise. Without promise pipelining this
would look something like this:

1.  Send message to create car factory
2.  Get back a reference to the car factory once it is made
3.  Send message to car factory to make a car
4.  Get back a reference to the car
5.  Send message to car to drive
6.  Get back the driving noise.

This is compared to when you use promise pipelining:

1.  Send message to create car factory
2.  Send message to car factory promise to make a car
3.  Send message to car promise to drive
4.  Get back driving noise

The other CapTP session, which is making the car factory, car, and finally car
noise, is able deliver those messages immediately to the local objects on their
side as they become fulfilled and send the car noise back at the end.

# [Cryptography](#cryptography)

> Each CapTP session has a per session key pair which is used for signing certain

Reviewer 2: Sometimes it is "key pair", like here. Other times it is "key-pair". Could one of those variants be used for consistency?

structures for example in the [third party handoffs
section](#third-party-handoffs). These key pair values are generated EdDSA with
a SHA512 hash.

**NOTE:** These representations are considered temporary and we are anticipating
replacing them, probably with record-based representations.

## [Public key](#public-key)

Public keys are formatted based on gcrypt's s-expression format, using EdDSA
public keys and the SHA512 hash algorithm. The EdDSA public keys are based on
the Ed25519 elliptic curve. The public key is formatted as follows:

```text
(public-key (ecc (curve Ed25519) (flags eddsa) (q ...) (s ...)))
```

Everything except the `...` must be verbatim. The `...` is replaced by Binary
Data typed values representing the public key.

## [Signature](#signature)

Signatures are formatted using gcrypt's s-expression format and the EdDSA
signature scheme. The formatted signature s-expression follows this structure:

```text
(sig-val (eddsa (r ...) (s ...)))
```

> Everything exception the `...` must be verbatim. The `...` is replaced by Binary

Reviewer: ```suggestion
Everything except the `...` must be verbatim. The `...` is replaced by Binary
```

Data typed values representing the signature.

# [Third Party Handoffs](#third-party-handoffs)

Third party handoffs are used when a message that is sent within a CapTP session
contains a reference to an object that has been imported from a different CapTP
session. This ability to include remote references to objects is a crucial part
of CapTP as it enables objects to use any references they hold in any context.

> The third party handoffs specified here ensures this can be done in a secure

Reviewer: ```suggestion
The third party handoffs specified here ensure this can be done in a secure
```

fashion, even if the messages are viewed by a malicious actor.

Third party handoffs define three roles:

-   **Gifter**: The session sharing their `desc:import-object` or
    `desc:import-promise`
-   **Receiver**: The session the `desc:import-object` or `desc:import-promise`
    is being shared with.
-   **Exporter**: The session exporting the `desc:import-object` or
    `desc:import-promise`

> ## Handoffs from the Gifter's prespective

Reviewer: ```suggestion
## Handoffs from the Gifter's perspective
```

succeed. The way to check which attempt should succeed is to sort the serialized
`session-pubkey`, the highest of the two wins.

### Constructing and sending

The `captp-version` MUST be `1.0`.

The `session-pubkey` is the public key part of the per-session key pair
generated for this connection. This is serialized in accordance with
[Cryptography](#cryptography)

The `acceptable-location` is a OCapN Locator which represents the location where
the session is accessable.

The `acceptable-location-sig` is the signature of the serialized
`acceptable-location`. The signature is created using the private key from the
per-session key pair. This is serialized in accordance with
[Cryptography](#cryptography)

### Receiving

The `captp-version` MUST be equal to `1.0`. If the version does not match, the
connection MUST be aborted.

The `acceptable-location-sig` MUST be valid that the `session-pubkey` provided a
valid signature of `acceptable-location`.

## [`op:bootstrap`](#op-bootstrap)

This operation requests the bootstrap object to be available at the specified
`answer-position`. The message looks like:

```text
<op:bootstrap [answer-position    ; positive integer or 0

>                resolve-me-desc]>  ; desc:import-object | desc:import-promise

Reviewer 5: resolve-me-desc here and in op:deliver seem to be redundant functionality of what op:listen does. See #15 for why.

-   `withdraw-gift`

#### `fetch` Method

This method is used to fetch an object from the bootstrap object. To use it you
need a `swiss-number` which is a Binary Data type. This swiss number should
correspond an object which exists in this session. The result will be the object
which corresponds to this `swiss-number` or an error if the object does not
exist or a swiss number was not provided.

An example of how to use this method is:

```text
<op:deliver <desc:export 0>          ; Remote bootstrap object
            ['fetch                  ; Argument 1: Symbol "fetch"
             swiss-number]           ; Argument 2: Binary Data
            3                        ; Answer position
            <desc:import-object 5>>  ; object exported by us at position 5 should provide the answer
```

#### `deposit-gift` Method

The deposit gift method is used in conjunction with sending a [Third Party
Handoff](#third-party-handoffs). This method is used to deposit a gift which has
been sent to the bootstrap object. It has two arguments:

1.  A gift ID that is positive integer.
2.  A `desc:import-object` or `desc:import-promise` which has been exported
    within the given CapTP session.

> 

Reviewer 5: missing an argument:
3. VatId of the recipiant of the gift.

Reviewer 5: In Spritely Goblins case here it would refer to the publiv key (or the cryptographics hash there of) of the endpont that is to get the gift.

@markm has the nifty phrase "My name is how you know it is me." which captures the essence of the idea of self authenticated designators.

Author: I don't think this is needed or adds anything to the handoff process.

The `desc:handoff-give` and the `desc:handoff-receive` already contain enough information that ensures that only the exporter can pick up the gift left for them.

Here is an example of how to use this method:

```text
<op:deliver-only <desc:export 0>                               ; Remote bootstrap object
                  ['deposit-gift                               ; Argument 1: Symbol "deposit-gift"
                  gift-id                                      ; Argument 2: Positive integer or 0
                  <desc:import-object | desc:import-promise>)> ; Argument 3
```

#### `withdraw-gift` Method

This method is used to send the [`desc:handoff-receive`](#desc-handoff-receive)
in order to receive a gift. It has one arguments:

- The `desc:handoff-receive`

This should have been sent with the `op:deliver` operation, the response the
bootstrap object should give is the gift which was (or will be) deposited.

> 

Reviewer 5: This actually needs two arguments:
1. gift id of the gift
2. VatId of the giver

             <desc:handoff-receive>]  ; Argument 2: desc:handoff-receive
            1                         ; Positive integer
            <desc:import-object 3>>   ; The object exported (by us) at position 3, should receive the gift. 
``` 

## [`op:deliver-only`](#op-deliver-only)

This operation delivers a message to an object and does not expect any result
from the receiving object.

The `op:deliver-only` message is:

```text
<op:deliver-only [to-desc   ; desc:export
                  args]>    ; Sequence
```

`to-desc` is a `desc:export` descriptor which corresponds to the object the
message is being sent to.

`args` is a sequence which are the arguments to invoke the object with.

Since this message has no result, it should be delivered to the object but
nothing further is required.

## [`op:deliver`](#op-deliver)

This operation delivers a message to an object with the expectation of a return
value which should be installed at the location specified in `answer-pos`. The
`op:deliver` message is:

``` text
<op:deliver [to-desc            ; desc:export
             args               ; sequence
             answer-pos         ; positive integer | false

>              resolve-me-desc]>  ; desc:import-object | desc:import-promise

Reviewer 5: resolve-me-desc here and in op:bootstrap seem to be redundant functionality of what op:listen does. See #15 for why.

Author: I actually hadn't seen #15 before so thank you for linking it. I think the concept of dropping the `resolve-me-desc` and then simplifying the `op:deliver` and `op:deliver-only` messages into a single `op:deliver` message and then relying on `op:listen`more is really interesting. I discussed it briefly with Christine and she thinks it's an interesting idea too.

I think this for now, this would be better addressed as a follow up PR if the group generally think this is a good idea (I think my initial impression is that I agree, this might be a nicer solution) we go ahead and make the change later.

This should not be referenced after the [`op:gc-answer`](#op-gc-answer) message
has been sent for this position.

```text
<desc:answer answer-pos> ; position: positive integer
```

## [`desc:sig-envelope`](#desc-sig-envelope)

This encapsulates a CapTP object and provides a signature. The signature is
created on the binary data of the serialized CapTP object in the `signed` field.

The process of generating this is:

1.  Fully serialize to Syrup octets a CapTP object.
2.  Sign the result of step 1 using the private key.
3.  Create a `desc:sig-envelope` with the CapTP object and signature.

```text
<desc:sig-envelope signed      ; captp-object
                   signature>  ; Signature (see cryptography section)
```

When this is received, the signature must be valid using the corresponding
public key. If the signature is not valid, the operation should be aborted.

## [`desc:handoff-give`](#desc-handoff-give)

You should have an overview of handoffs before reading this section, please read
the [Third Party Handoffs section](#third-party-handoffs) first.

This certificate is provided by the gifter to the exporter, the exporter then

> create a `desc:handoff-receive` which includes this `desc:handoff-give`

Reviewer 2: "then creates"

With the right tooling, many programming languages implementing CapTP can
achieve the same semantics of asynchronous programming for invoking both local
and remote objects, allowing programmers to focus on code flow rather than
networking infrastructure.

# Additional Documents

This document does not stand alone, it relies upon multiple other documents
which together build up OCapN (Object Capability Network) specifications.

This specification uses the following other specifications:

-   [Syrup](): The serialization format used for all messages between actors
    seperated by a CapTP boundary.
-   [OCapN Netlayers](): Specification to open a secure communication channel
    between two sessions, often on different networks.
-   [OCapN Locators](): Specification covers representation of object references
    for both in band and out of band usage.

# CapTP Overview

CapTP supports many features, as described in the [Introduction](#introduction)
section. It can, however, be broken down into the life cycle of a CapTP session.
A session is between two entities which are communicating via CapTP through a
OCapN Netlayer's channel. A CapTP life cycle consists of the following stages:

> 1. A session is established pairwise between machines.

Author: I've re-written this paragraph a little and moved away from life cycle closer to what you're suggesting here, let me know if this is what you had in mind.

from or because either side wishes to end the session. Both situations are
covered by the [`op:abort`](#op-abort) message. When this is received the
session should be severed and unresolved promises broken.

# Promises

Promises are a key part of CapTP. They are used to represent a value which is
not yet known. Promises are created by sending an `op:deliver` message, where a
promise is created for the value of the response. They can be chained together
in what is called [Promise Pipelining](#promise-pipelining), whereby messages
are sent to the promise which should be delivered if the promise resolves.

## [Promise and Resolver Objects](#promise-objects)

Promises work like regular objects in CapTP. Promises come as a pair, there is a
"promise" object which represents a value and may eventually resolve to either a
value or object reference or may break, either explicitly, implicitly through
error handling, or via network partition.

Messages can be sent to a promise as if it were the resolved object. These sent
messages will be relayed to the eventual object if it is `fulfill`ed to one.
There is also a "promise resolver" object which is used to provide the promise a

> value or break on an error. The resolver object has two methods:

Author: Goblins promises are built upon regular actors/objects and use op:deliver/op:deliver-only to function. This means that a promise is created via op:deliver (or in Goblins, you can also spawn them as you can with any other object, and share them within a message over the CapTP boundary). The promise object can have messages sent to it, and these will be forwarded to the object (or promise) it eventually resolves to, unless it breaks.

Goblins doesn't have a CapTP level op:resolve or anything similar; it simply has this resolver object which is spawned at the same time as the promise. When it's sent a message, it expects the first symbol to be 'fulfill' or 'break', followed by one or more values.

This is how promises work, both over CapTP and locally, and why this "promise API" documentation ends up in the specification. There are two such object APIs in the CapTP spec from Goblins' perspective: the bootstrap object with its three "methods", and the promise resolver with its two "methods".
